### ISSUE TITLE
faiss

### ISSUE THREAD

**Issue by jingweiz** [facebookresearch/XLM#133]
Hi,
Congrats on this super cool paper!
Might be a really stupid question, but just to make sure, in ```_get_indices()``` of ```HashingMemoryProduct```, ```faiss``` is used, while in ```HashingMemoryProductFast```, it is not used. Could you share a bit why not using ```faiss``` is faster?

**Comment by chqiwang** [comment]
I also want want to know

**Comment by glample** [comment]
Hi,

Don't use `HashingMemoryProduct`, use `HashingMemoryProductFast` (`--mem_implementation pq_fast`) it is faster both at training time and at inference. At inference you can make it even a bit faster by uncommenting https://github.com/facebookresearch/XLM/blob/master/src/model/memory/memory.py#L656-L657 and commenting the 4 lines above these 2.

**Comment by jingweiz** [comment]
That's very helpful, thank you so much for your response!
In the meanwhile could you share your go-to config when you train w/ the memory layer? Such as: ```--mem_query_detach_input --mem_query_layer_sizes --mem_query_kernel_sizes``` etc, that would be really helpful since it is not easy to figure out the best setting without your experience, thanks in advance!

**Comment by glample** [comment]
Hi,

I updated the memory with good parameters: https://github.com/facebookresearch/XLM/commit/f24f008b8e7e1de7e57f8e48e358afa606cd5dfe

It is also great if you can use batch normalization as suggested in the paper: `--mem_query_batchnorm 
 true` but I set the default value of this parameter to `False`, as it is not always possible to use it. Indeed, it requires the model to be trained with batches without padded sentences. If you have padding, the hidden states associated to the padding token will result in incorrect mean/variance in the batch norm layer. I will fix this later, but for now if you train with `--clm_step` or `--mlm_step` this won't be an issue (I added a check in the training script anyway to be sure this does not happen).

**Comment by jingweiz** [comment]
Thank you sooooooooooo much!

**Comment by andrew-sn** [comment]
> Hi,
> 
> Don't use `HashingMemoryProduct`, use `HashingMemoryProductFast` (`--mem_implementation pq_fast`) it is faster both at training time and at inference. At inference you can make it even a bit faster by uncommenting https://github.com/facebookresearch/XLM/blob/master/src/model/memory/memory.py#L656-L657 and commenting the 4 lines above these 2.

Hi, 

I am still confused about the difference between `HashingMemoryProductFast` and `HashingMemoryProduct`. It seems the difference with them is the former uses `F.linear` to get the scores and the latter uses `faiss`.

You recommend to use `HashingMemoryProductFast`, which is using `F.linear` to get the scores both at training and inference time, BUT also recommend to uncomment 2 lines(which leads to use `faiss` in `HashingMemoryProductFast`) to be faster as inference. Could you explain a bit detail?

**Comment by glample** [comment]
Those two implementations are a little bit different, it's not just about using FAISS or not.

The original idea in the "not fast" one, `HashingMemoryProduct`, was tu use FAISS to get the indices, then use these indices and `cartesian_product` to create an intermediate matrix of size `(bs, knn ** 2, 2)` to select the best indices overall.

However, FAISS does not work with gradient, it can only be used at inference, so idea was to then do this:
https://github.com/facebookresearch/XLM/blob/master/src/model/memory/memory.py#L605-L610
to recompute the scores with a gradient to the model parameters. In that setting, you are basically computing the scores twice, this is why it is not optimal.

Also, it turns out that these few lines of code above are extremely slow in PyTorch (see the comment line 605). I did not really understand why, probably some PyTorch optimization missing, but anyway both at inference and especially at training time the `fast` implementation is faster. Since there are rarely more than 1000 keys, you can safely ignore FAISS, searching over 1000 elements with FAISS or PyTorch does not make a big difference (performance becomes much better for FAISS if you search over 1M elements though, but with product keys you don't need that).
